### ISSUE TITLE
Refactor JSON Postgres tests

### ISSUE THREAD

**Issue by parsonsmatt** [yesodweb/persistent#1043]
The tests are currently written in a very direct style. They should be rewritten in `hspec`.

**Comment by JonathanLorimer** [comment]
I would be happy to take a run at this.

**Comment by parsonsmatt** [comment]
fantastic, thanks!

The trick is that the test suite mutates the state of a database, and runs tests against it as it changes state. So the hard part is ensuring that the tests have the same initial state as they do in the current suite, or rewriting the tests more invasively to have a different  prepare-act-assert mechanism.

**Comment by JonathanLorimer** [comment]
Hey Matt, I have been trying to take some time to figure how to approach this. I found this [library](https://hackage.haskell.org/package/hspec-pg-transact) which might be useful. I was wondering what your thoughts are on pulling this in?

**Comment by parsonsmatt** [comment]
That would be a little more invasive than I think we want to go with right now.

If you look at the `persistent=postgresql/test/ArrayAggTest.hs`, we have a convention, where the test suite looks like:

```haskell
specs :: RunDb SqlBackend IO -> Spec
  describe "rawSql/array_agg" $ do
    let runArrayAggTest :: (PersistField [a], Ord a, Show a) => Text -> [a] -> Assertion
        runArrayAggTest dbField expected = runDb $ do
          void $ insertMany
            [ UserPT "a" $ Just "b"
            , UserPT "c" $ Just "d"
            , UserPT "e"   Nothing
            , UserPT "g" $ Just "h" ]
          escape <- ((. DBName) . connEscapeName) `fmap` ask
          let query = T.concat [ "SELECT array_agg(", escape dbField, ") "
                               , "FROM ", escape "UserPT"
                               ]
          [Single xs] <- rawSql query []
          liftIO $ sort xs @?= expected

    it "works for [Text]"       $ do
        runArrayAggTest "ident"    ["a", "c", "e", "g" :: Text]
    it "works for [Maybe Text]" $ do
        runArrayAggTest "password" [Nothing, Just "b", Just "d", Just "h" :: Maybe Text]
```

Ideally, the jSON tests would look like:

```haskell
specs :: RunDb SqlBackend IO -> Spec
specs runDb = do
  describe "operator #1" $ do
    it "has a property" $ runDb $ do
      setup
      result <- use operator1 with setup data
      result `shouldBe` expectedValue
```

We can write `before` to perform an identical setup before each spec, which will probably be useful for many of the specs, since they share a setup.

**Comment by JonathanLorimer** [comment]
Okay, great, I will start there. Thanks Matt!

**Comment by JonathanLorimer** [comment]
Alright, I put up a [PR](https://github.com/yesodweb/persistent/pull/1072). I noticed a couple things while doing this refactor:

- I noticed that `db` is passed into most tests, but this function is available by simply by importing `PgInit`, would you like me to change it so that tests that import `PgInit` us `db` from there instead of receiving it as an argument?
- I was really confused at first by the difference between `db` and `runConn`, it took me a while to realize that `db` reverts all changes after it is run. Perhaps it would be helpful to rename it to something like `runConnRevert`? I am happy to make this change as well.

**Comment by parsonsmatt** [comment]
Great point - the `runDb` parameter is there to support running tests against multiple backends, so `MySQL` and `Postgresql` can share much of a test suite. But we don't *need* that in a postgres specific test. Nice.

`db` being renamed is probably a good call.

**Comment by parsonsmatt** [comment]
Closed by #1072
